compiler/optimizer/OMRValuePropagation.hpp

```diff
@@ -281,118 +284,32 @@ class ValuePropagation : public TR::Optimization
    ValueConstraint *copyValueConstraints(ValueConstraints &valueConstraints);
 
 #if USE_TREES
-   class ValueConstraintHandler : public TREE_HANDLER<ValueConstraint>
+   class ValueConstraintHandler : public TREE_HANDLER <ValueConstraint>
       {
       public:
-      void setVP(TR::ValuePropagation *vp) {_vp = vp;}
```

Comment:
I like this solution myself - I think it fixes all the problems: no need to cast, each layer is self contained (you want to add a constraint, a handler, and transformer for something in your language it stays in your language layer and the virtual inheritance makes it work), and best of all everyone else just include ValuePropagation.hpp and uses the TR::VP typedef.